File: main/sort-list/sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    static ListNode* sortList(ListNode* head) noexcept;
};

namespace {
    pair<ListNode*, ListNode*> merge(ListNode* first,
                                     ListNode* second) noexcept
    {
        ListNode sentinel {0};
        auto cur = &sentinel;

        for (; first && second; cur = cur->next) {
            if (second->val < first->val) {
                cur->next = second;
                second = second->next;
            } else {
                cur->next = first;
                first = first->next;
            }
        }

        cur->next = (first ? first : second);
        while (cur->next) cur = cur->next;

        return {sentinel.next, cur};
    }

    inline ListNode* advance(ListNode* node, unsigned distance) noexcept
    {
        while (node && distance--) node = node->next;
        return node;
    }

    ListNode* mergePass(ListNode* const head, const unsigned length) noexcept
    {
        assert(length != 0u);

        ListNode sentinel {0};
        sentinel.next = head;
        auto beforeFirst = &sentinel;

        for (; ; ) {
            const auto beforeSecond = advance(beforeFirst, length);
            const auto second = advance(beforeSecond, 1u);
            if (!second) break; // no more second halves to merge so we're done
            beforeSecond->next = nullptr;

            const auto beforeThird = advance(second, length - 1u);
            const auto third = advance(beforeThird, 1u);
            if (third) beforeThird->next = nullptr;

            ListNode* last {};
            tie(beforeFirst->next, last) = merge(beforeFirst->next, second);
            last->next = third;

            beforeFirst = last;
        }

        return beforeFirst == &sentinel ? nullptr : sentinel.next;
    }
}

ListNode* Solution::sortList(ListNode* head) noexcept
{
    for (auto length = 1u; ; length *= 2u) {
        const auto newHead = mergePass(head, length);
        if (!newHead) return head;
        head = newHead;
    }
}

File: main/sort-list/sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

import contextlib

class Solution:
    @staticmethod
    def sortList(head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        length = 1
        while True:
            newHead = Solution._mergePass(head, length)
            if newHead is None:
                break

            head = newHead
            length *= 2

        return head

    @staticmethod
    def _mergePass(head, length):
        sentinel = ListNode(...)
        sentinel.next = head
        beforeFirst = sentinel

        while True:
            beforeSecond = Solution._advance(beforeFirst, length)
            second = Solution._advance(beforeSecond, 1)
            if second is None:
                break # no second half to merge with
            beforeSecond.next = None

            beforeThird = Solution._advance(second, length - 1)
            third = Solution._advance(beforeThird, 1)
            if third is not None:
                beforeThird.next = None

            beforeFirst.next, last = Solution._merge(beforeFirst.next, second)
            last.next = third
            beforeFirst = last

        return None if beforeFirst is sentinel else sentinel.next

    @staticmethod
    def _advance(node, distance):
        with contextlib.suppress(AttributeError):
            for _ in range(distance):
                node = node.next
        return node

    @staticmethod
    def _merge(first, second):
        sentinel = ListNode(...)
        cur = sentinel

        while first is not None and second is not None:
            if second.val < first.val:
                cur.next = second
                second = second.next
            else:
                cur.next = first
                first = first.next
            cur = cur.next

        cur.next = (first if second is None else second)
        while cur.next is not None:
            cur = cur.next

        return sentinel.next, cur
